ksql-api/src/main/java/io/confluent/ksql/api/server/BlockingQueryPublisher.java

```diff
@@ -0,0 +1,175 @@
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A query publisher that uses an internal blocking queue to store rows for delivery. It's currently
+ * necessary to use a blocking queue as Kafka Streams delivers message in a synchronous fashion with
+ * no back pressure. If the queue was not blocking then if the subscriber was slow the messages
+ * could build up on the queue eventually resulting in out of memory. The only mechanism we have to
+ * slow streams down is to block the thread. Kafka Streams uses dedicated streams per topology so
+ * this won't prevent the thread from doing useful work elsewhere but it does mean we can't have too
+ * many push queries in the server at any one time as we can end up with a lot of threads. Ideally
+ * Kafka Streams would use a non-blocking reactive model with back-pressure, e.g. using reactive
+ * streams. That way a small number of threads would be required to service all topologies and we
+ * wouldn't need to block.
+ */
+public class BlockingQueryPublisher extends BasePublisher<GenericRow>
+    implements QueryPublisher, RowConsumer {
+
+  private static final Logger log = LoggerFactory.getLogger(BlockingQueryPublisher.class);
+
+  public static final int SEND_MAX_BATCH_SIZE = 10;
+  public static final int BLOCKING_QUEUE_CAPACITY = 1000;
+
+  private final BlockingQueue<GenericRow> queue = new LinkedBlockingQueue<>(
+      BLOCKING_QUEUE_CAPACITY);
+  private final WorkerExecutor workerExecutor;
+  private PushQueryHandler queryHandle;
+  private String[] columnNames;
+  private String[] columnTypes;
+  private OptionalInt limit;
+  private int numAccepted;
+  private boolean complete;
+  private volatile boolean closed;
+
+  public BlockingQueryPublisher(final Context ctx,
+      final WorkerExecutor workerExecutor) {
+    super(ctx);
+    this.workerExecutor = Objects.requireNonNull(workerExecutor);
+  }
+
+  public void setQueryHandle(final PushQueryHandler queryHandle) {
+    this.queryHandle = Objects.requireNonNull(queryHandle);
+    this.limit = queryHandle.getLimit();
+    this.columnNames = queryHandle.getColumnNames();
+    this.columnTypes = queryHandle.getColumnTypes();
+  }
+
+  @SuppressFBWarnings(value = "EI_EXPOSE_REP")
+  @Override
+  public String[] getColumnNames() {
+    return columnNames;
+  }
+
+  @SuppressFBWarnings(value = "EI_EXPOSE_REP")
+  @Override
+  public String[] getColumnTypes() {
+    return columnTypes;
+  }
+
+  public void close() {
+    if (closed) {
+      return;
+    }
+    closed = true;
+    // Run async as it can block
+    workerExecutor.executeBlocking(p -> queryHandle.stop(), ar -> {
+      if (ar.failed()) {
+        log.error("Failed to close query", ar.cause());
+      }
+    });
+    super.close();
+  }
+
+  @Override
+  public synchronized void accept(final GenericRow row) {
```

Comment:
accept could potentially be called by different threads, hence the synchronization

Reply:
This is still access state that `close()` accesses. is it possible that the two of those can be concurrent? If so we should also make that synchronized

Reply:
close() doesn't need synchronization as closed is volatile and workerExecutor is a final field

Reply:
I don't think volatile is sufficient in this context:
1. if two close() calls happen simultaneously, there's no guarantees that they both get past L89
2. if a close() and accept() happen at the same time, it is possible that the accept gets to L111 and then the close happens and the query handler stops.

I'm not sure how "bad" these race conditions are, but they are still races and I don't think there's a harm in proactively synchronizing (it's not that performance critical since accept is already synch'd and close isn't called often)